extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java

```diff
@@ -0,0 +1,259 @@
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.storage.s3.S3Utils;
+import org.apache.druid.storage.s3.ServerSideEncryptingAmazonS3;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class S3InputSource extends AbstractInputSource implements SplittableInputSource<URI>
+{
+  private static final Logger log = new Logger(S3InputSource.class);
+  private static final int MAX_LISTING_LENGTH = 1024;
+
+  private final ServerSideEncryptingAmazonS3 s3Client;
+  private final List<URI> uris;
+  private final List<URI> prefixes;
+  private Collection<URI> cacheSplitUris = null;
+
+  @JsonCreator
+  public S3InputSource(
+      @JacksonInject("s3Client") ServerSideEncryptingAmazonS3 s3Client,
+      @JsonProperty("uris") @Nullable List<URI> uris,
+      @JsonProperty("prefixes") @Nullable List<URI> prefixes
+  )
+  {
+    this.s3Client = Preconditions.checkNotNull(s3Client, "s3Client");
+    this.uris = uris == null ? new ArrayList<>() : uris;
+    this.prefixes = prefixes == null ? new ArrayList<>() : prefixes;
+
+    if (!this.uris.isEmpty() && !this.prefixes.isEmpty()) {
+      throw new IAE("uris and prefixes cannot be used together");
+    }
+
+    if (this.uris.isEmpty() && this.prefixes.isEmpty()) {
+      throw new IAE("uris or prefixes must be specified");
+    }
+
+    for (final URI inputURI : this.uris) {
+      Preconditions.checkArgument("s3".equals(inputURI.getScheme()), "input uri scheme == s3 (%s)", inputURI);
+    }
+
+    for (final URI inputURI : this.prefixes) {
+      Preconditions.checkArgument("s3".equals(inputURI.getScheme()), "input uri scheme == s3 (%s)", inputURI);
+    }
+  }
+
+  @JsonProperty
+  public List<URI> getUris()
+  {
+    return uris;
+  }
+
+  @JsonProperty("prefixes")
+  public List<URI> getPrefixes()
+  {
+    return prefixes;
+  }
+
+  @Override
+  public Stream<InputSplit<URI>> createSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec)
+      throws IOException
+  {
+    if (cacheSplitUris == null) {
+      initalizeSplitUris();
+    }
+    return cacheSplitUris.stream().map(InputSplit::new);
+  }
+
+  @Override
+  public int getNumSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec) throws IOException
+  {
+    if (cacheSplitUris == null) {
+      initalizeSplitUris();
+    }
+    return cacheSplitUris.size();
+  }
+
+  @Override
+  public SplittableInputSource<URI> withSplit(InputSplit<URI> split)
+  {
+    return new S3InputSource(s3Client, ImmutableList.of(split.get()), ImmutableList.of());
+  }
+
+  @Override
+  public boolean needsFormat()
+  {
+    return true;
+  }
+
+  @Override
+  protected InputSourceReader formattableReader(
```

Comment:
Unit test coverage is missing for this method

Reply:
added test through test of `read` method, which mocks the s3 client to do the list and getMetadata operations and mocked get object method that "returns" `S3Object` with csv file content